openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/streams/doc/man/man4/sc.4.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) sc.4.man,v 0.9.2.15 2006/04/05 09:43:45 brian Exp
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified 2006/04/05 09:43:45 by brian
.\"
.\" -------------------------------------------------------------------------
.\"
.\" sc.4.man,v
.\" Revision 0.9.2.15  2006/04/05 09:43:45  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database streams.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so streams.macros
.\"
.\"
.TH SC 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Modules"
.\"
.\"
.SH NAME
.B sc
\- \fISTREAMS\fR configuration module
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <sys/stropts.h>
.br
.B #include <sys/sc.h>
.HP
.IB t " = open(" /dev/nuls ", " 0 );
.HP
.BI ioctl( t ", " I_PUSH ", " \*(lqsc\*(rq );
.\"
.\"
.SH DESCRIPTION
.PP
.B sc
is a
.I STREAMS
configuration module for
.BR "Linux Fast-STREAMS" .
It interprets a set of
.B sc
.BR ioctl (2s)
commands and provides information about the streams configuration.
.B sc
does not interpret any
.BR M_DATA ", " M_PROTO " or " M_PCPROTO
messages either from above or below the module.
.B sc
can be pushed on the the null stream
.BR nuls (4)
or any other suitable stream available to the caller (e.g.
.BR sad (4)).
.\"
.\"
.SH USAGE
.PP
The
.B sc
module is not intended to be used directly by application program developers.
It is used by the
.BR scls (8)
utility an other utilities needing to enquire about the configuration of
.BR "Linux Fast-STREAMS" .
.\"
.\"
.SH IOCTLS
.PP
This subsection details the
.BR ioctl (2s)
commands that are made available by pushing the
.B sc
module.  For general information on input-output controls for
.I STREAMS
drivers, see
.BR streamio (7).
.\"
.SS SC_IOC_LIST
.PP
.I arg
is a pointer to a
.BR sc_list
structure in the user's address space.
.PP
The
.B sc_list
structure contains the following members:
.PP
.RS
.nf
\fC\s-1\
struct sc_list {
    int sc_nmods;
    struct sc_mlist *sc_mlist;
};
\s+1\fP
.fi
.TP \w'sc_mlist\(em'u
.I sc_nmods
On call specifies the number of
.B sc_mlist
structures that a provided in the caller's buffer pointed to by
.IR sc_mlist .
On return, specifies the number of
.B sc_mlist
structures that are available to be written to the caller's buffer pointed to
by
.IR sc_mlist ,
regardless of the space provided by the caller.
.TP
.I sc_mlist
Provides a pointer to a user buffer that contains at least
.I sc_nmods
number of
.B sc_mlist
structures in an array.
The module will returns nor more than the number of modules available, or the
number of modules requested,  whichever is less.  However,
.I sc_nmods
will always return the number of modules available.
.PP
By calling with a
.I sc_nmods
of zero and a
.I sc_mlist
of
.BR NULL ,
the caller can determine the number of modules that are available to
be returned in
.IR sc_mlist
so that the buffer may be appropriately sized for a subsequent call.
.RE
.PP
The
.B sc_mlist
structure contains the following members:
.PP
.RS
.nf
\fC\s-1\
struct sc_mlist {
    int major;
    struct module_info mi;
    struct module_stat ms;
};
\s+1\fP
.fi
.TP \w'major\(em'u
.I major
Specifies the
.I major
device number for the driver whose information is contained in this entry.
.TP
.I mi
Contains the
.B module_info
structure associated with the driver with the
.I major
device number specified.
Note that if the driver definition contains more than one
.BR module_info (9)
structure in the definition, only that associated with the write side of the upper multiplexer
(i.e.,
.I st_wrinit
in
.BR streamtab (9))
will be retrieved.
For details of the
.B module_info
structure, see
.BR module_info (9).
.TP
.I ms
Contains the
.B module_stats
structure associated with the driver with the
.I major
device number specified.
Note that if the driver definition contains more than one
.BR module_stat (9)
structure in the definition, only that associated with the write side of the upper multiplexer
(i.e.,
.I st_wrinit
in
.BR streamtab (9))
will be retrieved.
For details of the
.B module_stat
structure, see
.BR module_stat (9).
.RE
.\"
.\"
.SH NOTICES
.PP
Note that the
.B sc
module is not yet capable of copying and providing access to the private statistics area indicated
by the
.I ms_xprt
and
.I ms_xsize
members of the
.BR module_stat (9)
structure.  This is an unnecessary limitation.
At some point it would make sense to add a
.B SC_IOC_STATS
command that would retrieve both the
.BR module_stat (9)
structure and the private statistics for a single module or driver.
.\"
.\"
.SH EXAMPLES
.PP
See the source code for
.BR scls (8).
.\"
.\"
.SH MODULES
.PP
.B sc
is a
.I STREAMS
module named
.RI \*(lq sc \*(rq.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR ioctl (2s),
.BR nuls (4),
.BR sad (4)),
.BR scls (8),
.BR streamio (7),
.BR autopush (8),
.BR scls (8),
.BR streamtab (9),
.BR module_info (9),
.BR module_stat (9).
.\"
.\"
.SH BUGS
.PP
The
.B sc
module has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
Many implementations of
.I STREAMS
have some mechanism for augmenting the functionality of the
.BR sad (4)
driver.  Some implementations, such as
.IR AIX \(rg,
.[
aixspg
.]
.IR HP-UX \(rg,
.[
hpuxspg
.]
provide an
.B sc
module for this purpose.
.IP \(em \w'\(em\(em'u
.IR AIX \(rg
.[
aixspg
.]
provides an
.B sc
module for the purpose of allowing
.BR autopush (8)
to use device names instead of major device numbers, and to support the
.BR scls (8)
utility.
.IP \(em
.IR AIX \(rg
.[
aixspg
.]
requires the use of the
.B sc
modules with the
.BR nuls (4)
driver.
.IP \(em
.IR HP-UX \(rg
.[
hpuxspg
.]
describes the
.B sc
module as a core
.I STREAMS
module used by
.BR autopush (8)
and provides support for device names instead of device numbers, and providing
supplemental functions for the
.BR sad (4)
driver.
.IP \(em
.B sc
can be pushed over any stream, unlike the
.IR AIX \(rg
or
.IR HP-UX \(rg
documentation.
The user must have sufficient privilege to push the
.B sc
module.
.IP \(em
Internals of the
.B sc
module is implementation dependent, not subject to standardization, and not
documented.
This manual page documents the
.B "Linux Fast-STREAMS"
implementation of the
.B sc
module.
Compatibility with other implementations is not guaranteed.
.\"
.\"
.SH CONFORMANCE
.PP
None.  (This manual page.)
Conformance is verified using the
.BR test-sc (8)
verification test suite.
.\"
.\"
.SH HISTORY
.PP
An
.B sc
module for extending the capabilities of the
.BR sad (4)
driver appear in
.IR AIX \(rg
.[
aixspg
.]
and
.IR HP-UX \(rg
.[
hpuxspg
.]
documentation.
.\"
.\"
.[
$LIST$
.]
.TI


openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/streams/src/modules/sc.c
/*****************************************************************************

 @(#) sc.c,v OpenSS7-0_9_2_D_rc2(0.9.2.46) 2006/06/14 10:37:27

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; version 2 of the License.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified 2006/06/14 10:37:27 by brian

 -----------------------------------------------------------------------------

 sc.c,v
 Revision 0.9.2.46  2006/06/14 10:37:27  brian
 - defeat a lot of debug traces in debug mode for testing
 - changes to allow strinet to compile under LiS (why???)

 Revision 0.9.2.45  2006/03/10 07:23:58  brian
 - rationalized streams and strutil package sources

 Revision 0.9.2.44  2006/03/03 10:57:11  brian
 - 32-bit compatibility support, updates for release

 Revision 0.9.2.43  2006/02/20 10:59:24  brian
 - updated copyright headers on changed files

 *****************************************************************************/

#ident "@(#) sc.c,v OpenSS7-0_9_2_D_rc2(0.9.2.46) 2006/06/14 10:37:27"

static char const ident[] =
    "sc.c,v OpenSS7-0_9_2_D_rc2(0.9.2.46) 2006/06/14 10:37:27";

/* 
 *  This is SC, a STREAMS Configuration module for Linux Fast-STREAMS.  This
 *  is an auxilliary module to the SAD (STREAMS Administrative Driver) and can
 *  be pushed over that module or over the NULS (Null STREAM) driver.
 */

#include <linux/config.h>
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>

#include <sys/kmem.h>
#include <sys/cmn_err.h>
#include <sys/stream.h>
#include <sys/strconf.h>
#include <sys/strsubr.h>
#include <sys/ddi.h>

#include <sys/sc.h>

#include "sys/config.h"
#include "src/kernel/strlookup.h"

#define SC_DESCRIP	"UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define SC_COPYRIGHT	"Copyright (c) 1997-2006 OpenSS7 Corporation.  All Rights Reserved."
#define SC_REVISION	"LfS sc.c,v OpenSS7-0_9_2_D_rc2(0.9.2.46) 2006/06/14 10:37:27"
#define SC_DEVICE	"SVR 4.2 STREAMS STREAMS Configuration Module (SC)"
#define SC_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define SC_LICENSE	"GPL"
#define SC_BANNER	SC_DESCRIP	"\n" \
			SC_COPYRIGHT	"\n" \
			SC_REVISION	"\n" \
			SC_DEVICE	"\n" \
			SC_CONTACT	"\n"
#define SC_SPLASH	SC_DEVICE	" - " \
			SC_REVISION

#ifdef CONFIG_STREAMS_SC_MODULE
MODULE_AUTHOR(SC_CONTACT);
MODULE_DESCRIPTION(SC_DESCRIP);
MODULE_SUPPORTED_DEVICE(SC_DEVICE);
MODULE_LICENSE(SC_LICENSE);
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-sc");
#endif
#endif

#ifndef CONFIG_STREAMS_SC_NAME
#error CONFIG_STREAMS_SC_NAME must be defined.
#endif
#ifndef CONFIG_STREAMS_SC_MODID
#error CONFIG_STREAMS_SC_MODID must be defined.
#endif

modID_t modid = CONFIG_STREAMS_SC_MODID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0);
#endif
MODULE_PARM_DESC(modid, "Module ID for SC.");

#ifdef MODULE_ALIAS
#ifdef LFS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_SC_MODID));
MODULE_ALIAS("streams-module-sc");
#endif
#endif

static struct module_info sc_minfo = {
	.mi_idnum = CONFIG_STREAMS_SC_MODID,
	.mi_idname = CONFIG_STREAMS_SC_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

#if defined __LP64__ && defined LFS
#  undef WITH_32BIT_CONVERSION
#  define WITH_32BIT_CONVERSION 1
#endif

static size_t
sc_mlist_copy(long major, struct qinit *qinit, caddr_t _mlist, const uint flag)
{
#ifdef WITH_32BIT_CONVERSION
	if (flag == IOC_ILP32) {
		struct sc_mlist32 *mlist = (typeof(mlist)) _mlist;

		if ((mlist->major = major) != -1) {
			strncpy(mlist->name, qinit->qi_minfo->mi_idname, FMNAMESZ + 1);
			// mlist->mi = *qinit->qi_minfo;
			mlist->mi.mi_idnum = qinit->qi_minfo->mi_idnum;
			strncpy(mlist->mi.mi_idname, qinit->qi_minfo->mi_idname, FMNAMESZ + 1);
			mlist->mi.mi_minpsz = qinit->qi_minfo->mi_minpsz;
			mlist->mi.mi_maxpsz = qinit->qi_minfo->mi_maxpsz;
			mlist->mi.mi_hiwat = qinit->qi_minfo->mi_hiwat;
			mlist->mi.mi_lowat = qinit->qi_minfo->mi_lowat;
			if (qinit->qi_mstat) {
				mlist->ms.ms_pcnt = qinit->qi_mstat->ms_pcnt;
				mlist->ms.ms_scnt = qinit->qi_mstat->ms_scnt;
				mlist->ms.ms_ocnt = qinit->qi_mstat->ms_ocnt;
				mlist->ms.ms_ccnt = qinit->qi_mstat->ms_ccnt;
				mlist->ms.ms_acnt = qinit->qi_mstat->ms_acnt;
				mlist->ms.ms_flags = qinit->qi_mstat->ms_flags;
				mlist->ms.ms_xsize = qinit->qi_mstat->ms_xsize;
			}
		}
		return sizeof(struct sc_mlist32);
	} else
#endif				/* WITH_32BIT_CONVERSION */
	{
		struct sc_mlist *mlist = (typeof(mlist)) _mlist;

		if ((mlist->major = major) != -1) {
			strncpy(mlist->name, qinit->qi_minfo->mi_idname, FMNAMESZ + 1);
			// mlist->mi = *qinit->qi_minfo;
			mlist->mi.mi_idnum = qinit->qi_minfo->mi_idnum;
			strncpy(mlist->mi.mi_idname, qinit->qi_minfo->mi_idname, FMNAMESZ + 1);
			mlist->mi.mi_minpsz = qinit->qi_minfo->mi_minpsz;
			mlist->mi.mi_maxpsz = qinit->qi_minfo->mi_maxpsz;
			mlist->mi.mi_hiwat = qinit->qi_minfo->mi_hiwat;
			mlist->mi.mi_lowat = qinit->qi_minfo->mi_lowat;
			if (qinit->qi_mstat) {
				mlist->ms.ms_pcnt = qinit->qi_mstat->ms_pcnt;
				mlist->ms.ms_scnt = qinit->qi_mstat->ms_scnt;
				mlist->ms.ms_ocnt = qinit->qi_mstat->ms_ocnt;
				mlist->ms.ms_ccnt = qinit->qi_mstat->ms_ccnt;
				mlist->ms.ms_acnt = qinit->qi_mstat->ms_acnt;
				mlist->ms.ms_flags = qinit->qi_mstat->ms_flags;
				mlist->ms.ms_xsize = qinit->qi_mstat->ms_xsize;
			}
		}
		return sizeof(struct sc_mlist);
	}
}

static size_t
str_mlist_count(void)
{
	return (cdev_count + fmod_count);
}

/* 
 *  -------------------------------------------------------------------------
 *
 *  PUT routine
 *  
 *  -------------------------------------------------------------------------
 */
static streamscall int
sc_wput(queue_t *q, mblk_t *mp)
{
	union ioctypes *ioc;
	int err = 0, rval = 0;
	mblk_t *dp = mp->b_cont;

	switch (mp->b_datap->db_type) {
	case M_FLUSH:
		/* canonical flushing */
		if (mp->b_rptr[0] & FLUSHW) {
			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(q, mp->b_rptr[1], FLUSHALL);
			else
				flushq(q, FLUSHALL);
			mp->b_rptr[0] &= ~FLUSHW;
		}
		if (mp->b_rptr[0] & FLUSHR) {
			queue_t *rq = RD(q);

			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(rq, mp->b_rptr[1], FLUSHALL);
			else
				flushq(rq, FLUSHALL);
			qreply(q, mp);
			return (0);
		}
		break;
	case M_IOCTL:
	{
		caddr_t uaddr;
		size_t usize;

		_trace();
		ioc = (typeof(ioc)) mp->b_rptr;

#ifdef WITH_32BIT_CONVERSION
		if (ioc->iocblk.ioc_flag == IOC_ILP32) {
			uaddr = (caddr_t) (unsigned long) (uint32_t) *(unsigned long *) dp->b_rptr;
			usize = sizeof(struct sc_list32);
		} else
#endif
		{
			uaddr = (caddr_t) *(unsigned long *) dp->b_rptr;
			usize = sizeof(struct sc_list);
		}
		switch (ioc->iocblk.ioc_cmd) {
		case SC_IOC_LIST:
			/* there is really no reason why a regular user cannot list modules and
			   related information. */
#if 0
			_trace();
			err = -EPERM;
			if (ioc->iocblk.ioc_uid != 0) {
				_ptrace(("Error path taken!\n"));
				goto nak;
			}
#endif
			_trace();
			if (ioc->iocblk.ioc_count == TRANSPARENT) {
				_trace();
				if (uaddr == NULL) {
					rval = str_mlist_count();
					goto ack;
				}
				mp->b_datap->db_type = M_COPYIN;
				ioc->copyreq.cq_addr = uaddr;
				ioc->copyreq.cq_size = usize;
				ioc->copyreq.cq_flag = 0;
				ioc->copyreq.cq_private = (mblk_t *) 0;
				qreply(q, mp);
				return (0);
			}
			_trace();
			/* doesn't support I_STR yet, just TRANSPARENT */
			err = -EINVAL;
			goto nak;
		}
		break;
	}
	case M_IOCDATA:
		ioc = (typeof(ioc)) mp->b_rptr;

		switch (ioc->copyresp.cp_cmd) {
		case SC_IOC_LIST:
			_trace();
			if (ioc->copyresp.cp_rval != 0) {
				_ptrace(("Aborting ioctl!\n"));
				goto abort;
			}
			_trace();
			if (ioc->copyresp.cp_private == (mblk_t *) 0) {
				int n, count;
				caddr_t uaddr;
				size_t usize;

				n = 0;
				if (!dp || dp->b_wptr == dp->b_rptr) {
					rval = str_mlist_count();
					goto ack;
				}
#ifdef WITH_32BIT_CONVERSION
				if (ioc->copyresp.cp_flag == IOC_ILP32) {
					if (dp->b_wptr < dp->b_rptr + sizeof(struct sc_list32)) {
						_ptrace(("Error path taken!\n"));
						err = -EFAULT;
						goto nak;
					} else {
						struct sc_list32 *sclp = (typeof(sclp)) dp->b_rptr;

						count = sclp->sc_nmods;
						uaddr = (caddr_t) (unsigned long) sclp->sc_mlist;
						usize = count * sizeof(struct sc_mlist32);
					}
				} else
#endif
				{
					if (dp->b_wptr < dp->b_rptr + sizeof(struct sc_list)) {
						_ptrace(("Error path taken!\n"));
						err = -EFAULT;
						goto nak;
					} else {
						struct sc_list *sclp = (typeof(sclp)) dp->b_rptr;

						count = sclp->sc_nmods;
						uaddr = (caddr_t) sclp->sc_mlist;
						usize = count * sizeof(struct sc_mlist);
					}
				}
				if (count < 0) {
					_ptrace(("Error path taken!\n"));
					err = -EINVAL;
					goto nak;
				}
				if (count > 100) {
					_ptrace(("Error path taken!\n"));
					err = -ERANGE;
					goto nak;
				}
				if (count == 0) {
					rval = str_mlist_count();
					goto ack;
				}
				if (!(dp = allocb(usize, BPRI_MED))) {
					_ptrace(("Error path taken!\n"));
					err = -ENOSR;
					goto nak;
				}
				dp->b_wptr = dp->b_rptr + usize;
				bzero(dp->b_rptr, usize);
				freemsg(mp->b_cont);
				mp->b_cont = dp;
				{
					struct list_head *pos;
					uint flag = ioc->copyresp.cp_flag;
					caddr_t mlist = (typeof(mlist)) dp->b_rptr;

					_trace();
					if (n < count) {
						_trace();
						/* output all devices */
						read_lock(&cdevsw_lock);
						list_for_each(pos, &cdevsw_list) {
							struct cdevsw *cdev;
							struct qinit *qinit;

							if (n >= count)
								break;
							cdev =
							    list_entry(pos, struct cdevsw, d_list);
							qinit = cdev->d_str->st_wrinit;
							mlist +=
							    sc_mlist_copy(cdev->d_major, qinit,
									  mlist, flag);
							n++;
						}
						read_unlock(&cdevsw_lock);
					}
					_trace();
					if (n < count) {
						_trace();
						/* output all modules */
						read_lock(&fmodsw_lock);
						list_for_each(pos, &fmodsw_list) {
							struct fmodsw *fmod;
							struct qinit *qinit;

							if (n >= count)
								break;
							fmod =
							    list_entry(pos, struct fmodsw, f_list);
							qinit = fmod->f_str->st_wrinit;
							mlist +=
							    sc_mlist_copy(0, qinit, mlist, flag);
							n++;
						}
						read_unlock(&fmodsw_lock);
					}
					_trace();
					/* zero all excess elements */
					for (; n < count; n++) {
						mlist += sc_mlist_copy(-1, NULL, mlist, flag);
					}
				}
				_trace();
				mp->b_datap->db_type = M_COPYOUT;
				ioc->copyreq.cq_addr = uaddr;
				ioc->copyreq.cq_size = usize;
				ioc->copyreq.cq_flag = 0;
				ioc->copyreq.cq_private = (mblk_t *) (long) count;
				qreply(q, mp);
				return (0);
			} else {
				_trace();
				/* done */
				rval = (int) (long) ioc->copyresp.cp_private;
				goto ack;
			}
		}
	      nak:
		mp->b_datap->db_type = M_IOCNAK;
		ioc->iocblk.ioc_count = 0;
		ioc->iocblk.ioc_rval = -1;
		ioc->iocblk.ioc_error = -err;
		qreply(q, mp);
		return (0);
	      ack:
		mp->b_datap->db_type = M_IOCACK;
		ioc->iocblk.ioc_count = 0;
		ioc->iocblk.ioc_rval = rval;
		ioc->iocblk.ioc_error = 0;
		qreply(q, mp);
		return (0);
	      abort:
		_ctrace(freemsg(mp));
		return (0);
	}
	putnext(q, mp);
	return (0);
}
static streamscall int
sc_rput(queue_t *q, mblk_t *mp)
{
	putnext(q, mp);
	return (0);
}

/* 
 *  -------------------------------------------------------------------------
 *
 *  OPEN and CLOSE
 *
 *  -------------------------------------------------------------------------
 */
static streamscall int
sc_open(queue_t *q, dev_t *devp, int oflag, int sflag, cred_t *crp)
{
	if (q->q_ptr)
		return (0);	/* already open */
	if (sflag == MODOPEN && WR(q)->q_next != NULL) {
		/* don't need to be privileged to push the module, just to use it */
#if 0
		if (crp->cr_uid != 0 && crp->cr_ruid != 0)
			return (-EACCES);
#endif
		q->q_ptr = WR(q)->q_ptr = (void *) 1;
		qprocson(q);
		return (0);
	}
	return (-EIO);		/* can't be opened as driver */
}
static streamscall int
sc_close(queue_t *q, int oflag, cred_t *crp)
{
	qprocsoff(q);
	q->q_ptr = WR(q)->q_ptr = NULL;
	return (0);
}

/* 
 *  -------------------------------------------------------------------------
 *
 *  Registration and initialization
 *
 *  -------------------------------------------------------------------------
 */

static struct qinit sc_rqinit = {
	.qi_putp = sc_rput,
	.qi_qopen = sc_open,
	.qi_qclose = sc_close,
	.qi_minfo = &sc_minfo,
};

static struct qinit sc_wqinit = {
	.qi_putp = sc_wput,
	.qi_minfo = &sc_minfo,
};

static struct streamtab sc_info = {
	.st_rdinit = &sc_rqinit,
	.st_wrinit = &sc_wqinit,
};

static struct fmodsw sc_fmod = {
	.f_name = CONFIG_STREAMS_SC_NAME,
	.f_str = &sc_info,
	.f_flag = D_MP,
	.f_kmod = THIS_MODULE,
};

static void *sc_opaque;

static void
sc_unregister_ioctl32(void)
{
	if (sc_opaque != NULL)
		unregister_ioctl32(sc_opaque);
}

static int
sc_register_ioctl32(void)
{
	if ((sc_opaque = register_ioctl32(SC_IOC_LIST)) == NULL)
		return (-ENOMEM);
	return (0);
}

#ifdef CONFIG_STREAMS_SC_MODULE
static
#endif
int __init
sc_init(void)
{
	int err;

#ifdef CONFIG_STREAMS_SC_MODULE
	printk(KERN_INFO SC_BANNER);
#else
	printk(KERN_INFO SC_SPLASH);
#endif
	sc_minfo.mi_idnum = modid;
	if ((err = sc_register_ioctl32()) < 0)
		return (err);
	if ((err = register_strmod(&sc_fmod)) < 0) {
		sc_unregister_ioctl32();
		return (err);
	}
	if (modid == 0 && err > 0)
		modid = err;
	return (0);
};

#ifdef CONFIG_STREAMS_SC_MODULE
static
#endif
void __exit
sc_exit(void)
{
	unregister_strmod(&sc_fmod);
	sc_unregister_ioctl32();
};

#ifdef CONFIG_STREAMS_SC_MODULE
module_init(sc_init);
module_exit(sc_exit);
#endif
